src/rgw/rgw_motr.h

```diff
@@ -0,0 +1,70 @@
+#ifndef CEPH_RGWMOTR_H
+#define CEPH_RGWMOTR_H
+#include <functional>
+#include <boost/container/flat_map.hpp>
+
+#include "include/rados/librados.hpp"
+#include "include/Context.h"
+#include "include/random.h"
+#include "common/RefCountedObj.h"
+#include "common/ceph_time.h"
+#include "common/Timer.h"
+#include "rgw_common.h"
+#include "cls/rgw/cls_rgw_types.h"
+#include "cls/version/cls_version_types.h"
+#include "cls/log/cls_log_types.h"
+#include "cls/timeindex/cls_timeindex_types.h"
+#include "cls/otp/cls_otp_types.h"
+#include "rgw_log.h"
+#include "rgw_metadata.h"
+#include "rgw_meta_sync_status.h"
+#include "rgw_period_puller.h"
+#include "rgw_obj_manifest.h"
+#include "rgw_sync_module.h"
+#include "rgw_trim_bilog.h"
+#include "rgw_service.h"
+#include "rgw_sal.h"
+#include "rgw_aio.h"
+#include "rgw_d3n_cacherequest.h"
+
+#include "services/svc_rados.h"
+#include "services/svc_bi_rados.h"
+#include "common/Throttle.h"
+#include "common/ceph_mutex.h"
+#include "rgw_cache.h"
+
+
+class RGWMotr
+{
+protected:
+  CephContext *cct;
+ public:
+  RGWMotr():cct(NULL),
+               pctl(&ctl)
+                {} 
+  std::string host_id ="";
+  int initialize(CephContext *_cct, const DoutPrefixProvider *dpp) {
+    set_context(_cct);
+    return initialize(dpp);
+  }
+    void set_context(CephContext *_cct) {
+    cct = _cct;
+  }
+  int initialize(const DoutPrefixProvider *dpp);
+  RGWServices svc;
```

Comment:
This data structure contains many members which are defined in services/, those classes/structures defined in that directory are implemented specific to RADOS only. And those classes also stores metadata in RADOS's system objects. Simply including RGWServices here is not a good idea as it potentially will run into lots of issues as when configuring with Motr backend, there is no RADOS backend to store those services metadata. 

Have you ran any test on your patch? even luckily the patch doesn't show any issue for now, IMO, the method is not good.

A temporary solution is to implement those functions (get_host_id(), zone_unique_id() etc) ourself and avoid using RGWServices.

A better solution is to generalise the system object implementation (services/*) to make it support multiple backends, but it will take a great deal of effort and better do it with RGW community.

Reply:
I agree with @siningwuseagate. We better avoid using RGWServices from motr sal code.

Reply:
Yes.  These "services" are, as I understand it, specific to RADOS and broadly, the notion of generalizing "system object" definitely needs design discussion.  Such as, could we get rid of it.

Reply:
Added this to our next meeting agenda at https://pad.ceph.com/p/rgw-weekly.